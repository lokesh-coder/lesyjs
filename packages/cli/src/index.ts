const commands = [`${__dirname}/commands`];
const plugins = [
  "@lesy/lesy-plugin-generator",
  "@lesy/lesy-plugin-prompt",
  "@lesy/lesy-plugin-help",
  [
    "@lesy/lesy-plugin-pilot",
    { docTitle: "Lesy | Dashboard", appName: "Lesy Dashboard" },
  ],
];

export { commands, plugins };
