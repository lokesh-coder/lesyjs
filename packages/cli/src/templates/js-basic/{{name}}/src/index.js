const commands = [`${__dirname}/commands`];

module.exports = { commands };
