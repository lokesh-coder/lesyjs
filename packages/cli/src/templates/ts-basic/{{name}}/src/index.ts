const commands = [`${__dirname}/commands`];

export { commands };
