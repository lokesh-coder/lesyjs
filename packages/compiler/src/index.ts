import { existsSync } from "fs";

class LesyCompiler {
  opts: any = {
    isTypescriptApp: false,
    loadDefaultPlugins: true,
    srcFilePath: "src/index.ts",
    tsFlag: "--ts",
    root: null,
    customTsConfig: null,
  };

  private defaultPlugins = ["@lesy/lesy-plugin-sidekick"];
  private defaultConfig = {
    version: { enable: true, flags: ["--version", "-v"] },
    defaultCommand: "default",
  };
  private tsConfigPath = null;

  exec(opts = {}) {
    this.opts = { ...this.opts, ...opts };
    const {
      isTypescriptApp,
      root,
      srcFilePath,
      tsFlag,
      loadDefaultPlugins,
      customTsConfig,
      ...src
    } = this.opts;
    const hasSrcData = Object.keys(src).length > 0;
    const hasBuildDir = existsSync(`${root}/dist`);
    const shouldRunTSCode = process.argv.includes(tsFlag);
    let mainFilePath: string;
    process.env.LESY_LANG = "js";
    if (!isTypescriptApp) {
      mainFilePath = !hasSrcData && require.resolve(root);
    } else {
      if ((hasBuildDir && shouldRunTSCode) || !hasBuildDir) {
        // run src
        process.env.LESY_LANG = "ts";
        mainFilePath = `${root}/${srcFilePath}`;
        this.tsConfigPath = customTsConfig || `${root}/tsconfig.json`;
        this.registerTSNode();
        this.registerPaths();
      } else {
        // run dist
        mainFilePath = !hasSrcData && require.resolve(root);
      }
    }

    return this.runApp(hasSrcData ? src : require(mainFilePath));
  }

  private registerTSNode() {
    const tsNode = require("ts-node");
    tsNode.register({
      transpileOnly: true,
      typeCheck: false,
      project: this.tsConfigPath,
    });
  }

  private registerPaths() {
    if (!require(this.tsConfigPath)["compilerOptions"]["paths"]) {
      return;
    }
    const tsPaths = require("tsconfig-paths");
    const config = tsPaths.loadConfig(this.tsConfigPath); // todo: duplicate
    tsPaths.register({ baseUrl: config.absoluteBaseUrl, paths: config.paths });
  }

  private runApp(appData) {
    const { LesyCore, LesyCoreClass } = require("@lesy/core");
    const { root } = this.opts;
    Object.assign(appData, this.includeEssentials(appData));

    const lesy = LesyCore.bootstrap({ ...appData, root }).catch((e: any) =>
      console.log("LESY ERROR:", e),
    );

    return {
      parse: (argv: string[]) =>
        lesy.then((l: typeof LesyCoreClass) =>
          l.run(argv || process.argv.slice(2)),
        ),
    };
  }

  private includeEssentials(appData) {
    let { plugins = [], config = {} } = appData;
    if (this.opts.loadDefaultPlugins) {
      plugins = [
        ...this.defaultPlugins.map((p: string) => require.resolve(p)),
        ...plugins,
      ];
    }
    config = { ...this.defaultConfig, ...config };
    return { ...appData, plugins, config };
  }
}

const compiler = new LesyCompiler();
module.exports = compiler.exec.bind(compiler);
