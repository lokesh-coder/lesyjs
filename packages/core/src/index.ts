export { LesyCore, LesyCoreClass } from "./core";
export * from "./model";
