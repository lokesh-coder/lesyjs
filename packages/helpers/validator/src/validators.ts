export const requiredValidator = value => !!value;
export const exactValidator = (value, rule) => value === rule;
export const sizeValidator = (value, rule) => {
  return !!value && value.length >= rule.min && value.length <= rule.max;
};
