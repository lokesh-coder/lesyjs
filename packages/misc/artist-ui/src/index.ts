import { store, onChange } from "@fabiospampinato/store";
import { parse as htmlparse } from "himalaya";
import logUpdate from "log-update";
import { minify } from "html-minifier";
import { flattenToStr } from "./utils";
import * as uiElements from "./elements";

interface Config {
  collapseWhitespace?: boolean;
  keepClosingSlash?: boolean;
  preserveLineBreaks?: boolean;
}

interface ElementDef {
  name: string;
  render: Function;
  onInit?: Function;
}

interface Element {
  [key: string]: ElementDef;
}

interface ElProp {
  key: string;
  value: string;
}

const defaultConfig = {
  collapseWhitespace: false,
  keepClosingSlash: true,
  preserveLineBreaks: false,
};

export default class Artist {
  private store: { [key: string]: any };
  private config: Config = defaultConfig;
  private onInitHook: Function = () => {};
  private elements = {};
  private timers = {
    internal: {},
    global: {},
  };
  private timerIDs = {
    internal: {},
    global: {},
  };
  private runningElTimers = [];

  constructor(config: Config = {}) {
    this.store = store({});
    this.config = { ...this.config, ...config };
    this.registerEls(uiElements);
  }

  registerEls(elements: Element) {
    const allElements = { ...this.elements, ...elements };
    Object.keys(allElements).forEach((key) => {
      const { name, ...data } = allElements[key];
      this.elements[name] = data;
    });
  }

  private runElHook(hook: Function, props: { [key: string]: any }) {
    if (!hook) return;
    hook({
      props,
      store: this.store,
      timer: this.registerInternalTimer.bind(this),
    });
  }

  private registerGlobalTimer(
    timerFunc: Function,
    interval: number,
    name: string,
  ) {
    this.timerIDs.global[name] = setInterval(timerFunc, interval);
    this.timers.global[name] = true;
  }

  private registerInternalTimer(
    timerFunc: Function,
    interval: number,
    name: string,
  ) {
    this.runningElTimers.push(name);
    if (this.timers.internal[name]) {
      return;
    }

    this.timerIDs.internal[name] = setInterval(timerFunc, interval);
    this.timers.internal[name] = true;
  }

  private clearUnusedTimers() {
    const allTimers = Object.keys(this.timers.internal);
    const runningElTimers = this.runningElTimers;
    const unusedTimers = allTimers.reduce((acc, item) => {
      if (!runningElTimers.includes(item)) acc.push(item);
      return acc;
    }, []);

    unusedTimers.forEach((name: string) => this.disposeTimer(name, "internal"));
    this.runningElTimers = [];
  }

  clearAllTimers() {
    Object.keys(this.timers.internal).forEach((name: string) =>
      this.disposeTimer(name, "internal"),
    );
    Object.keys(this.timers.global).forEach((name: string) =>
      this.disposeTimer(name, "global"),
    );
    this.runningElTimers = [];
  }

  disposeTimer(name: string, scope: string) {
    clearInterval(this.timerIDs[scope][name]);
    delete this.timers[scope][name];
  }

  private getProps(propsArr: ElProp[]) {
    return propsArr.reduce(
      (acc, prop) => Object.assign(acc, { [prop.key]: prop.value }),
      {},
    );
  }

  private renderEl(elements: any[]) {
    const output = [];
    elements.forEach((el) => {
      if (Array.isArray(el)) {
        output.push(this.renderEl(el));
        return;
      }
      const { attributes, tagName, type, content } = el;
      const props = this.getProps(attributes || []);
      const ctx = {
        tagName,
        props,
        renderEl: this.renderEl.bind(this),
        store: this.store,
      };

      if (type === "text" && content !== "") {
        const textEl = this.elements["text"];
        this.runElHook(textEl.init, props);
        const textRes = textEl.render(ctx, el);
        if (textRes) output.push(textRes);
      } else if (tagName) {
        const elFnDef = this.elements[tagName] || this.elements["unknown"];
        this.runElHook(elFnDef?.init, props);
        output.push(this.renderEl([htmlparse(elFnDef.render(ctx, el))]));
      }
    });

    return flattenToStr(output).join("");
  }

  rewriteScreen(output: string) {
    logUpdate(`[2K${output}`);
  }

  private updateScreen(tplFn: Function, init = true) {
    const tpl = tplFn(this.store);
    const tplsminified = minify(tpl, this.config);
    const elements = htmlparse(tplsminified);
    const output = this.renderEl(elements);
    if (!init) this.clearUnusedTimers();
    this.rewriteScreen(output);
  }

  onInit(fn: Function) {
    this.onInitHook = fn;
  }

  render(tplFn: Function) {
    this.onInitHook(this.store, this.registerGlobalTimer.bind(this));
    this.updateScreen(tplFn);
    onChange(this.store, () => {
      this.updateScreen(tplFn, false);
    });
  }
}

// const customElem = {
//   name: "custom",
//   render: () => `hello # <spinner type="moon"/> #`,
// };

// // USAGE

// const artist = new Artist();

// artist.registerEls({ customElem });

// let showSpinner = true;
// let showSpinner2 = false;

// artist.onInit((store, timer) => {
//   if (!store.count) store.count = 1;
//   timer(
//     () => {
//       store.count += 1;
//     },
//     1000,
//     "hello",
//   );

//   setTimeout(() => {
//     showSpinner = false;
//     showSpinner2 = true;
//   }, 1000);
// });

// artist.render((store) => {
//   return `
//   my name is <custom/> <text>hi</text> and ${store.count}

//   another spinner2: ${showSpinner2 && "<spinner/>"}

//   <text bold color='red'>another</text> spinner: ${showSpinner && "<spinner/>"}
//   // im a commebt
// /* me to */
// <box>
// <progress score="20"/>
//   </box>

// <text color="green"><row border>
//     <column>one</column>
//     <column>two</column>
//     <column>3</column>
//     </row></text>

// <div>hello</div><div>how are you</div>
// <span>hello</span><span>how are you</span>
// <div>you are <space length="300"/> yes <newline/> awesome </text>
// `;
// });

// setTimeout(() => {
//   artist.clearAllTimers();
// }, 2000);
