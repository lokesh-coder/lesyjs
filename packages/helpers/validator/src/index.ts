import { exactValidator, requiredValidator, sizeValidator } from "./validators";

export class LesyValidator {
  private validators = {};
  private status: any[] = [];
  constructor(private schema: object) {
    this.register("exact", exactValidator);
    this.register("required", requiredValidator);
    this.register("size", sizeValidator);
  }
  async validate(obj: object) {
    if (!obj) return [];
    const entries = Object.entries(this.schema);
    for (let [name, rules] of entries) {
      if (typeof rules !== "object") {
        throw new Error(`invalid rules provide in ${name}`);
      }

      const errors = {};
      rules = Object.entries(rules).filter(([name, rule]) => {
        errors[name] = rule;
        return !name.includes("Error");
      });

      for (const [n, rule] of rules) {
        if (!this.validators[n]) continue;

        const isPassed = await this.validators[n](obj[name], rule);
        const error =
          (!isPassed &&
            (errors[`${n}Error`] ||
              `<${n}> validation failed for [${name}]`)) ||
          null;

        this.status.push({
          name,
          validator: n,
          passed: isPassed,
          error: typeof error === "function" ? error(obj[name]) : error,
        });
      }
    }

    return this.status;
  }
  register(
    name: string,
    fn: (val, rule) => Boolean | Promise<Boolean>,
    aliases: any[] = [],
  ): any {
    this.validators[name] = fn;
    aliases.forEach((a: any) => {
      this.validators[a] = fn;
    });
  }
}

// let va2 = new LesyValidator(['name[*,3<->7,==foobar|abc,/^abc/]','age']);
