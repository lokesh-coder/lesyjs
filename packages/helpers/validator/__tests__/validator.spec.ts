import { LesyValidator } from "../src/index";

describe("LesyValidator", () => {
  afterEach(() => {
    expect.hasAssertions();
  });

  it("should succeed if schema is empty", async () => {
    const lesyValidator = new LesyValidator({});
    expect(await lesyValidator.validate({ name: "test" })).toEqual([]);
  });
  it("should throw err if required param is falsy", async () => {
    const lesyValidator = new LesyValidator({ name: { required: true } });
    const validation = await lesyValidator.validate({});

    expect(validation).toMatchSnapshot();
    expect(validation.length).toEqual(1);
    expect(validation[0].passes).toBeFalsy();
  });
  it("should succeed if required param is truthy", async () => {
    const lesyValidator = new LesyValidator({ name: { required: true } });
    const validation = await lesyValidator.validate({ name: "test" });

    expect(validation).toMatchSnapshot();
    expect(validation.length).toEqual(1);
    expect(validation[0].passed).toBeTruthy();
  });
  it("should succeed if no validators defined in schema", async () => {
    const lesyValidator = new LesyValidator({});
    const validation = await lesyValidator.validate({ name: "test" });

    expect(validation.length).toEqual(0);
  });
  it("should throw errs as array if multiple props failed", async () => {
    const lesyValidator = new LesyValidator({
      name: { required: true },
      age: { required: true },
    });
    const validation = await lesyValidator.validate({ name: null });

    expect(validation).toMatchSnapshot();
    expect(validation.length).toEqual(2);
    expect(validation[0].passed).toBeFalsy();
    expect(validation[1].passed).toBeFalsy();
  });
  it("should throw errs as array if multiple validators failed", async () => {
    const lesyValidator = new LesyValidator({
      name: { required: true, foo: "bar", size: { min: 4, max: 10 } },
      age: { required: true },
    });
    const validation = await lesyValidator.validate({ name: null });

    expect(validation.length).toEqual(3);
    expect(validation.every((v: any) => v.passed)).toBeFalsy();
  });
  it("should throw err if schema is invalid", async () => {
    const lesyValidator = new LesyValidator({
      name: "some text",
    });
    try {
      await lesyValidator.validate({ name: "test" });
    } catch (e) {
      expect(e.message).toEqual("invalid rules provide in name");
    }
  });
  it("should register custom validator", async () => {
    const lesyValidator = new LesyValidator({
      name: { enum: ["hello", "world"] },
    });
    lesyValidator.register("enum", (value, rule) => {
      return rule.includes(value);
    });

    const validation = await lesyValidator.validate({ name: "hello" });

    expect(validation.length).toEqual(1);
    expect(validation[0].passed).toBeTruthy();
  });
  it("should succeed for async validator", async () => {
    const lesyValidator = new LesyValidator({
      name: { delay: 100 },
    });
    lesyValidator.register(
      "delay",
      (_, __): Promise<Boolean> => {
        return new Promise((res, rej) => {
          res(false);
        });
      },
    );

    const validation = await lesyValidator.validate({ name: "hello" });
    expect(validation.length).toEqual(1);
    expect(validation[0].passed).toBeFalsy();
  });
  it("should have errs as array if multiple mixed validators failed ", async () => {
    const lesyValidator = new LesyValidator({
      name: { size: { min: 8, max: 10 }, delay: 100 },
    });
    lesyValidator.register(
      "delay",
      (_, __): Promise<Boolean> => {
        return new Promise((res, rej) => {
          res(false);
        });
      },
    );

    const validation = await lesyValidator.validate({ name: "hello" });
    expect(validation.length).toEqual(2);
    expect(validation.every((v: any) => v.passed)).toBeFalsy();
  });
  it("should have err in the same order in schema for sync validators", async () => {
    const lesyValidator = new LesyValidator({
      name: { required: true, size: { min: 4, max: 10 } },
    });

    const validation = await lesyValidator.validate({ name: null });

    expect(validation.length).toEqual(2);
    expect(validation[0].passed).toBeFalsy();
    expect(validation.every((v: any) => v.passed)).toBeFalsy();
    expect(validation.map((v: any) => v.validator)).toEqual(
      expect.arrayContaining(["required", "size"]),
    );
  });
  it("should have default err when validation failed", async () => {
    const lesyValidator = new LesyValidator({
      name: { required: true },
    });

    const validation = await lesyValidator.validate({});

    expect(validation.length).toEqual(1);
    expect(validation[0].error).toEqual(
      "<required> validation failed for [name]",
    );
  });
  it("should have err with custom message", async () => {
    const lesyValidator = new LesyValidator({
      name: { required: true, requiredError: "value is required!" },
    });

    const validation = await lesyValidator.validate({});

    expect(validation.length).toEqual(1);
    expect(validation[0].error).toEqual("value is required!");
  });
  it("should have errs when custom err message func is available", async () => {
    const lesyValidator = new LesyValidator({
      name: {
        required: true,
        requiredError: "value is required!",
        size: { min: 4, max: 10 },
        sizeError: (value: string) => {
          return `size of ${value} doesnt meet the required length`;
        },
      },
    });

    const validation = await lesyValidator.validate({ name: "hi" });

    expect(validation.length).toEqual(2);
    expect(validation.map((v: any) => v.error)).toEqual([
      null,
      "size of hi doesnt meet the required length",
    ]);
  });
  it("should able to override inbuild validators", async () => {
    const lesyValidator = new LesyValidator({
      name: { required: true },
    });
    lesyValidator.register("required", (value, rule) => {
      return false;
    });

    const validation = await lesyValidator.validate({ name: "hello" });

    expect(validation.length).toEqual(1);
    expect(validation[0].passed).toBeFalsy();
  });
  it("should accept aliases for validator name", async () => {
    const lesyValidator = new LesyValidator({ name: { noSpace: true } });

    lesyValidator.register("empty", (val, _) => !val.includes(" "), [
      "noSpace",
    ]);

    const validation = await lesyValidator.validate({ name: "hello buddy" });

    expect(validation.length).toEqual(1);
    expect(validation[0].passed).toBeFalsy();
  });
  it("should ignore if subject prop is not object", async () => {
    const lesyValidator = new LesyValidator({ name: { required: true } });
    const validation = await lesyValidator.validate(null as any);

    expect(validation.length).toEqual(0);
  });
  it("should ignore unknown rules", async () => {
    const schema = { name: { required: true, description: "user name" } };
    const obj = { name: "John" };

    const lesyValidator = new LesyValidator(schema);
    const validation = await lesyValidator.validate(obj);

    expect(validation.length).toEqual(1);
  });
});
