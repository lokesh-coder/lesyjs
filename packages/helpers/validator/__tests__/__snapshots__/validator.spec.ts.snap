// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LesyValidator should succeed if required param is truthy 1`] = `
Array [
  Object {
    "error": null,
    "name": "name",
    "passed": true,
    "validator": "required",
  },
]
`;

exports[`LesyValidator should throw err if required param is falsy 1`] = `
Array [
  Object {
    "error": "<required> validation failed for [name]",
    "name": "name",
    "passed": false,
    "validator": "required",
  },
]
`;

exports[`LesyValidator should throw errs as array if multiple props failed 1`] = `
Array [
  Object {
    "error": "<required> validation failed for [name]",
    "name": "name",
    "passed": false,
    "validator": "required",
  },
  Object {
    "error": "<required> validation failed for [age]",
    "name": "age",
    "passed": false,
    "validator": "required",
  },
]
`;
